p2p/host/resource-manager/conn_limiter.go

```diff
@@ -0,0 +1,138 @@
+package rcmgr
+
+import (
+	"net/netip"
+	"sync"
+)
+
+type ConnLimitPerCIDR struct {
+	// How many leading 1 bits in the mask
+	BitMask   int
+	ConnCount int
+}
+
+// 8 for now so that it matches the number of concurrent dials we may do
+// in swarm_dial.go. With future smart dialing work we should bring this
+// down
+var defaultMaxConcurrentConns = 8
+
+var defaultIP4Limit = ConnLimitPerCIDR{
+	ConnCount: defaultMaxConcurrentConns,
+	BitMask:   32,
+}
+var defaultIP6Limits = []ConnLimitPerCIDR{
+	{
+		ConnCount: defaultMaxConcurrentConns,
+		BitMask:   56,
+	},
+	{
+		ConnCount: 8 * defaultMaxConcurrentConns,
+		BitMask:   48,
+	},
+}
+
+func WithLimitPeersPerCIDR(ipv4 []ConnLimitPerCIDR, ipv6 []ConnLimitPerCIDR) Option {
+	return func(rm *resourceManager) error {
+		if ipv4 != nil {
+			rm.connLimiter.connLimitPerCIDRIP4 = ipv4
+		}
+		if ipv6 != nil {
+			rm.connLimiter.connLimitPerCIDRIP6 = ipv6
+		}
+		return nil
+	}
+}
+
+type connLimiter struct {
+	mu                  sync.Mutex
+	connLimitPerCIDRIP4 []ConnLimitPerCIDR
+	connLimitPerCIDRIP6 []ConnLimitPerCIDR
+	ip4connsPerLimit    []map[string]int
+	ip6connsPerLimit    []map[string]int
```

Comment:
Why is this a slice? How are these maps ever garbage collected?

Reply:
It’s a slice because the limits are a slice. Each map corresponds to each limit. 

You’re right about the gc of course, I need to delete after removing all conns from a subnet.